scarlet/component.py

```diff
@@ -70,64 +75,37 @@ def parameters(self):
         return [p for p in self._parameters if not p.fixed]
 
     @abstractmethod
-    def get_model(self, *parameters):
+    def get_model(self, *parameters, frame=None):
         """Get the model for this component
 
         Parameters
         ----------
         parameters: tuple of optimimzation parameters
 
+        frame: `~scarlet.frame.Frame`
+            Frame to project the model into. If `frame` is `None`
+            then the model contained in `bbox` is returned.
+
         Returns
         -------
         model: array
             (Channels, Height, Width) image of the model
         """
         pass
 
-    def set_frame(self, frame):
+    def set_model_frame(self, model_frame):
         """Sets the frame for this component.
 
-        Each component needs to know the properties of the Frame and, potentially, the
-        subvolume it covers.
+        Each component needs to know the properties of the Frame and,
+        potentially, the subvolume it covers.
 
         Parameters
         ----------
-        frame: `~scarlet.Frame`
-            Frame to adopt for this component
+        model_frame: `~scarlet.Frame`
+            Frame of the model
         """
-        self.frame = frame
-
-        # store padding and slicing structures
-        if self.bbox is not None:
-            assert isinstance(self.bbox, Box)
-            # TODO: full 3D bbox and slicing support
-            # determine pad from box into full frame
-            # yields superset of frame pixels
-            # pad_width is ((before1, after1), (before2, after2)...)
-            self.pad_width = list(
-                (
-                    max(0, self.bbox.start[d] - self.frame.start[d]),
-                    max(0, self.frame.stop[d] - self.bbox.stop[d]),
-                )
-                for d in range(self.frame.D)
-            )
-
-            # get slicing of padded box so that the result covers
-            # all of the model frame
-            bounds = []
-            for d in range(self.frame.D):
-                bounds.append(
-                    (
-                        self.bbox.start[d] - self.pad_width[d][0],
-                        self.bbox.stop[d] + self.pad_width[d][1],
-                    )
-                )
-            padded_box = Box.from_bounds(*bounds)
-
-            model_box = self.frame
-            overlap = padded_box & model_box
-            overlap -= padded_box.origin  # now in padded frame
-            self.slices = overlap.slices_for(padded_box.shape)
+        self.model_frame = model_frame
+        self.model_frame_slices, self.model_slices = overlapped_slices(model_frame, self.bbox)
```

Comment:
Really: that's it? I worked hard for the code that you now replace with merely 2 lines...